src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs

```diff
@@ -615,14 +614,10 @@ private static SuggestedActionSetPriority GetSuggestedActionSetPriority(CodeActi
                     supportsFeatureService.SupportsRefactorings(document) &&
                     requestedActionCategories.Contains(PredefinedSuggestedActionCategoryNames.Refactoring))
                 {
-                    // It may seem strange that we kick off a task, but then immediately 'Wait' on 
-                    // it. However, it's deliberate.  We want to make sure that the code runs on 
-                    // the background so that no one takes an accidentally dependency on running on 
-                    // the UI thread.
-                    var refactorings = Task.Run(
-                        () => _owner._codeRefactoringService.GetRefactoringsAsync(
-                            document, selection, cancellationToken),
-                        cancellationToken).WaitAndGetResult(cancellationToken);
+                    // Note that the refactorings are now allowed to run on UI thread.
+                    var refactorings = ThreadingContext.JoinableTaskFactory.Run(
+                        async () => await _owner._codeRefactoringService.GetRefactoringsAsync(
+                            document, selection, cancellationToken));
```

Comment:
that's pretty concerning.  that means a refactoring could block user input.  We've worked *hard* to ensure that thsi is never necessary.  is there a reason we must have a refactoring run on the UI thread

Reply:
Yes, I agree. But it seems to be the general consensus of the team is that we have to do this sooner or later. @tmat and @heejaechang can provide more context. I made the change here because the value of rootnamespace defined in C# project is not available from compiler (it doesn't need it at all), so I have to use `IProjectManagementService.GetDefaultNamespace` which requires UI thread.

Reply:
we decide to embrace JTF. and whether something uses JTF or not is not leaked (caller can't know whether callee uses JTF to go to UI thread) outside and that's by design of JTF. so we need to change all Task.Run/Task.Wait to JTF version.